### PR TITLE
ERC-7504 parity

### PR DESCRIPTION
**Pull request by nkrishang**

### CODE REVIEW

File: src/interface/IExtensionManager.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     *  @notice Add a single function to an existing extension.
+     */
+    function addFunctionToExtension(string memory extensionName, ExtensionFunction memory extFunction) external;
+    
+    /**
+     *  @notice Remove a single function from an existing extension.
+     */
+    function removeFunctionFromExtension(string memory extensionName, bytes4 functionSelector) external;
```
Comment (joaquim-verges) [inline src/interface/IExtensionManager.sol]: you didn't go with enable/disable in the end here?

File: src/lib/ExtensionManagerStorage.sol [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+pragma solidity ^0.8.0;
+
+import "./StringSet.sol";
+import "../interface/IExtension.sol";
+
+library ExtensionManagerStorage {
+
+    /// @custom:storage-location erc7201:extension.manager.storage
+    bytes32 public constant EXTENSION_MANAGER_STORAGE_POSITION = keccak256(abi.encode(uint256(keccak256("extension.manager.storage")) - 1));
```
Comment (joaquim-verges) [inline src/lib/ExtensionManagerStorage.sol:12]: erc7201 niiice

File: src/example/RouterImmutable.sol [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        return false;
+    }
+
+    /// @dev Returns whether a function can be added to an extension in the given execution context.
+    function _canAddFunctionToExtension(string memory, ExtensionFunction memory) internal view virtual override returns (bool) {
+        return false;
+    }
+
+    /// @dev Returns whether an extension can be removed from an extension in the given execution context.
+    function _canRemoveFunctionFromExtension(string memory, bytes4) internal view virtual override returns (bool) {
```
Comment (joaquim-verges) [inline src/example/RouterImmutable.sol]: do we  really need a permission function for each action?

File: src/interface/IExtensionManager.sol (hunk 1 above)
Comment (nkrishang) [reply src/interface/IExtensionManager.sol]: Yes, I'll be renaming:

- `addExtensionToFunction` -> `enableFunctionInExtension`
- `removeExtensionFromFunction` -> `disableFunctionInExtension`

File: src/example/RouterImmutable.sol (hunk 3 above)
Comment (nkrishang) [reply src/example/RouterImmutable.sol]: I've written `_can` functions for each function and placed each function's relevant require checks in them. I like the pattern, and doesn't limit the `_can` functions to permissions. What do you think?

Comment (joaquim-verges) [reply src/example/RouterImmutable.sol]: it just feels heavy to have to override all these _can functions when most of the time you'd have the same logic in them no?

Comment (nkrishang) [reply src/example/RouterImmutable.sol]: Makes sense. Added an `isAuthorizedCallToUpgrade` function to `ExtensionManager`. You can check the `/example` contracts for the usage of `isAuthorizedCallToUpgrade` and `_can...` functions.

**Pull request merged**